File: web/src/dashboard/mapTimeline.ts
import type { Bug, Goal, Scenario, Score, Task, TestRun } from "./types";
import { deriveScenarioState, type ScenarioState } from "./scenarioState";
import { buildMap, type MapGraph } from "./mapView";

/** Normalize a Firestore Timestamp / number / absent into millis (null when unknown). */
export function tsMillis(v: unknown): number | null {
  if (typeof v === "number") return v;
  if (v && typeof (v as { toMillis?: () => number }).toMillis === "function") {
    return (v as { toMillis: () => number }).toMillis();
  }
  return null;
}

/** One loop's run data (loopId undefined = project-direct "main"). */
export interface LoopSlice {
  loopId?: string;
  tasks: Task[];
  bugs: Bug[];
  scores: Score[];
  testRuns: TestRun[];
}

/** Missing createdAt (legacy data) ⇒ treated as always-present, keeping growth monotonic. */
const within = (cutoff: number) => (e: { createdAt?: unknown }) => {
  const t = tsMillis(e.createdAt);
  return t === null || t <= cutoff;
};

/** Cross-loop ids can collide (each loop names its own tasks/bugs); scope merged ids by loop. */
const scoped = (loopId: string | undefined, id: string) => (loopId ? `${loopId}.${id}` : id);

/** The graph as of time T: entities filtered to createdAt <= T; scenario met-state evaluated
 *  over only the events with createdAt <= T (deriveScenarioState already picks the
 *  latest-by-ULID, so restricting its input arrays restricts it to "latest within cutoff").
 *  Bugs render while open at T (created <= T and not yet fixed at T) — the one sanctioned
 *  exception to monotonic growth, mirroring the live view's open-bugs-only rule. */
export function mapAtTime(input: { goals: Goal[]; scenarios: Scenario[]; slices: LoopSlice[]; cutoff: number }): MapGraph {
  const { goals, scenarios, slices, cutoff } = input;
  const inWindow = within(cutoff);

  const goalsT = goals.filter(inWindow);
  const scenariosT = scenarios.filter(inWindow);
  const scoresT = slices.flatMap((sl) => sl.scores).filter(inWindow);
  const runsT = slices.flatMap((sl) => sl.testRuns).filter(inWindow);

  const scenarioStates: Record<string, ScenarioState> = {};
  for (const s of scenariosT) scenarioStates[s.id] = deriveScenarioState(s, scoresT, runsT);

  const tasksT: Task[] = slices.flatMap((sl) =>
    sl.tasks.filter(inWindow).map((t) => ({ ...t, id: scoped(sl.loopId, t.id), loopId: sl.loopId })));
  const openAtT = (b: Bug) => {
    const fixed = tsMillis(b.fixedAt);
    return inWindow(b) && (b.status !== "fixed" || fixed === null || fixed > cutoff);
  };
  const bugsT: Bug[] = slices.flatMap((sl) =>
    sl.bugs.filter(openAtT).map((b) => ({
      ...b,
      id: scoped(sl.loopId, b.id),
      taskId: b.taskId ? scoped(sl.loopId, b.taskId) : b.taskId, // bug→task edges stay within the loop
      loopId: sl.loopId,
    })));

  return buildMap({ goals: goalsT, scenarios: scenariosT, scenarioStates, tasks: tasksT, currentTaskId: null, openBugs: bugsT });
}

File: ios/Autoloop/Domain/MapTimeline.swift
import Foundation

/// Growth-replay: the product map as of time T — mirrors web mapTimeline.ts.

/// One loop's run data (loopId nil = project-direct "main").
struct MapSlice {
    var loopId: String? = nil
    var tasks: [MapTask] = []
    var bugs: [MapBug] = []
    var scores: [ScoreRec] = []     // carry createdAt
    var testRuns: [TestRunRec] = [] // carry createdAt
}

/// Missing createdAt (legacy data) ⇒ treated as always-present, keeping growth monotonic.
private func within(_ cutoff: Date, _ createdAt: Date?) -> Bool {
    guard let t = tsMillis(createdAt) else { return true }
    return t <= cutoff.timeIntervalSince1970 * 1000
}

/// Cross-loop ids can collide (each loop names its own tasks/bugs); scope merged ids by loop.
private func scoped(_ loopId: String?, _ id: String) -> String { loopId.map { "\($0).\(id)" } ?? id }

/// The graph as of time T: entities filtered to createdAt <= T; scenario met-state evaluated over
/// only the events with createdAt <= T. Bugs render while open at T (created <= T and not yet fixed
/// at T) — the one sanctioned exception to monotonic growth, mirroring the live open-bugs rule.
func mapAtTime(goals: [MapGoal], scenarios: [MapScenario], slices: [MapSlice], cutoff: Date) -> MapGraph {
    let goalsT = goals.filter { within(cutoff, $0.createdAt) }
    let scenariosT = scenarios.filter { within(cutoff, $0.createdAt) }
    let scoresT = slices.flatMap { $0.scores }.filter { within(cutoff, $0.createdAt) }
    let runsT = slices.flatMap { $0.testRuns }.filter { within(cutoff, $0.createdAt) }

    var scenarioStates: [String: MapNodeState] = [:]
    for s in scenariosT {
        let st = deriveScenarioState(ScenarioRec(id: s.id, threshold: s.threshold), scores: scoresT, testRuns: runsT)
        scenarioStates[s.id] = st.state == .met ? .met : .unmet
    }

    let tasksT: [MapTask] = slices.flatMap { sl in
        sl.tasks.filter { within(cutoff, $0.createdAt) }.map { t in
            var t = t; t.id = scoped(sl.loopId, t.id); t.loopId = sl.loopId; return t
        }
    }
    func openAtT(_ b: MapBug) -> Bool {
        guard within(cutoff, b.createdAt) else { return false }
        if b.status != "fixed" { return true }
        guard let fixed = tsMillis(b.fixedAt) else { return true }
        return fixed > cutoff.timeIntervalSince1970 * 1000
    }
    let bugsT: [MapBug] = slices.flatMap { sl in
        sl.bugs.filter(openAtT).map { b in
            var b = b
            b.id = scoped(sl.loopId, b.id)
            if let tid = b.taskId { b.taskId = scoped(sl.loopId, tid) }
            b.loopId = sl.loopId
            return b
        }
    }

    return buildMap(goals: goalsT, scenarios: scenariosT, scenarioStates: scenarioStates,
                    tasks: tasksT, currentTaskId: nil, openBugs: bugsT)
}
